shared/route-tree/index.js

```diff
@@ -256,18 +256,19 @@ export function routeSetState(
   routeDef: ?RouteDefNode,
   routeState: ?RouteStateNode,
   path: Path,
-  partialState: {}
+  partialState: {} | ((oldState: I.Map<string, any>) => I.Map<string, any>)
 ): RouteStateNode {
   const pathSeq = I.Seq(path)
   const name = pathSeq.first()
-  if (!name) {
-    // $FlowIssue
-    return routeState.update('state', state => state.merge(partialState))
+  if (!name && routeState) {
+    return typeof partialState === 'function'
```

Comment:
this is new, you can pass a state update function to setRouteState for atomic updates